tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,192 @@
+
+        # re-caculate the network interrupts distribution
+        # the increased NIC IRQs should be distributed in CPU cores in balance
+        my @total_irqs_on_cpu_after_network_download = get_nic_irqs_distribution_from_guest($guest, $nproc);
+        my @increased_irqs_on_cpu;
+        for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
+            $increased_irqs_on_cpu[$cpu_id] = $total_irqs_on_cpu_after_network_download[$cpu_id] - $initial_total_irqs_on_cpu[$cpu_id];
+            die("The value of network interrupts from download for CPU" . $cpu_id . " is 0") if $increased_irqs_on_cpu[$cpu_id] == 0;
+        }
+        record_info("IRQs of downloads distribution", "@increased_irqs_on_cpu");
+    }
+
+    restore_original_guests();
+
+}
+
+#set up guest test environment to run irqbalance test
+sub prepare_guest_for_irqbalance {
+    my $vm_name = shift;
+
+    #4 or more vcpu is needed
+    my $nproc = script_output "virsh vcpucount $vm_name | grep current | grep config | awk '{ print \$3 }'";
+    if ($nproc < 4) {
+        assert_script_run "virsh dumpxml --inactive $vm_name > $vm_name.xml";
+        script_run "virsh destroy $vm_name";
+        assert_script_run "xmlstarlet edit -L -d /domain/vcpu $vm_name.xml";
+        assert_script_run "xmlstarlet edit -L \\
+                               -s /domain -t elem -n vcpu -v 4 \\
+			       -s /domain/vcpu -t attr -n placement -v static \\
+			       $vm_name.xml";
+        script_run "virsh undefine $vm_name" unless (script_run "virsh undefine --nvram $vm_name") == 0;
+        assert_script_run " ! virsh list --all | grep $vm_name";
+        assert_script_run "virsh define $vm_name.xml";
+        assert_script_run "virsh start $vm_name";
+    }
+
+    wait_guest_online($vm_name);
+    assert_script_run "ssh root\@$vm_name \"zypper in -y irqbalance\"" unless script_run("ssh root\@$vm_name \"rpm -q irqbalance\"") eq 0;
+
+}
+
+# return the total NIC IRQs on each CPU core from guest
+sub get_nic_irqs_distribution_from_guest {
+    my ($vm_name, $cpu_account) = @_;
+
+    my @total_irqs_on_cpu = ();
+    my $irq_output = script_output("ssh root\@$vm_name \"grep vif0 /proc/interrupts\"");
+    record_info("NIC IRQs distribution", $irq_output);
+    for (my $cpu_id = 0; $cpu_id < $cpu_account; $cpu_id++) {
+        my @irqs_on_one_cpu = split('\n', script_output("ssh root\@$vm_name \"awk '/vif0/{ print \\\$(($cpu_id+2)) }' /proc/interrupts\""));
+        # sum of IRQs on one cpu core
+        for (my $i = 0; $i <= $#irqs_on_one_cpu; $i++) {
+            $total_irqs_on_cpu[$cpu_id] += $irqs_on_one_cpu[$i];
+        }
+    }
+
+    return @total_irqs_on_cpu;
+}
+
+# it is more stable to get the affinities after a little time than just after it is written
+# especially with irqbalance is enabled, so a function is wrapped
+sub get_irq_affinities_from_guest {
+    my ($vm_name, @irqs) = @_;
+
+    my $cat_affinity_files_cmdline = "ssh root\@$vm_name \"cd /proc/irq && cat";
```

Comment:
why do not get smp_affinity file via cat tool directly
For example:
`"ssh root\@$vm_name \"cat /proc/irq/"; `

Reply:
I aim to get the file content of /proc/irq/<NIC_IRQ_NUMBER>/smp_affinity. so the actural command finally is gotten run is `"ssh root@guest_name \"cd /proc/irq && cat 83/smp_affinit 84/smp_affinity 85/smp_affinity ... 92/smp_affinity \""`